### PR TITLE
Add tracing spans for rules API

### LINKED ISSUE #4131: Add tracing for federated rules API
### Describe the solution you'd like

Instrument tracing spans for the federated APIs. This should include the proxy (querier) side as well as the leaf store sides.

### PR DESCRIPTION
**Pull request by toshi1127**
Signed-off-by: Matsumoto <toshi.matsumoto.3n@gmail.com>

* [ ] I added CHANGELOG entry for this change.
* [x] Change is not relevant to the end user.

## Changes

Add tracing for federated rules API.

Closes #4131 

## Verification

- `make build`
- `make test-local`

### CODE REVIEW

**Review by kakkoyun — DISMISSED** [review]
LGTM 🥇 Thanks for your contribution.
Great, clean PR.

**Review by yeya24 — COMMENTED** [review]
Overall it looks good. But one thing I am not sure about is that it is too fine-grained, which might create a lot of spans for a single Rules request. 
I am not very familiar with trace instrumentation, is there a best practice about it?

File: pkg/rules/manager.go [hunk 1]
```diff
@@ -135,6 +136,10 @@ func NewManager(
 		extLset:   extLset,
 		ruleFiles: make(map[string]string),
 	}
+
+	span, ctx := tracing.StartSpan(ctx, "new_manager")
+	defer span.Finish()
```
Comment (yeya24) [inline pkg/rules/manager.go:141]: I think this is not needed.

Comment (toshi1127) [reply pkg/rules/manager.go]: Thanks for the review!
I've finished fixing it.

**Comment by toshi1127** [comment]
I don't know the best practices either, but...
I think adding trace spans for each element might be useful to determine unnecessary trace spans in the future.

File: pkg/rules/manager.go [hunk 2]
```diff
@@ -391,6 +392,7 @@ func (m *Manager) Update(evalInterval time.Duration, files []string) error {
 
 // Rules returns specified rules from manager. This is used by gRPC and locally for HTTP and UI purposes.
 func (m *Manager) Rules(r *rulespb.RulesRequest, s rulespb.Rules_RulesServer) error {
+	var err error
```
Comment (bwplotka) [inline pkg/rules/manager.go:395]:
```suggestion
func (m *Manager) Rules(r *rulespb.RulesRequest, s rulespb.Rules_RulesServer) (err error) {
```

We can simplify (:

**Review by bwplotka — DISMISSED** [review]
Thanks a lot! Just one nit (: 

I think this amount is fine.

**Comment by yeya24** [comment]
Hi @toshi1127, please rebase on main then we can get CI fixed.

**Review by yeya24 — APPROVED** [review]
LGTM!

**Review by kakkoyun — APPROVED** [review]

**Pull request merged**